Fix ValidationAwareTrait relying on Table const

lilHermit opened the pull request:
as discussed in #11106

lilHermit committed:
Fix ValidationAwareTrait relying on Table const

robertpustulka commented on src/Validation/ValidatorAwareInterface.php:
```diff
@@ -21,6 +21,13 @@
+     *
+     * @var string
+     */
+    const BUILD_VALIDATOR_EVENT = 'Model.buildValidator';
```
I don't think this can be `Model.buildValidator` as the trait can be used outside model scope.

dereuromark replied:
But is it then still BC if you change the value?

robertpustulka replied:
You can keep the old value in `Table` class and add some generic name in the interface. The constant will get overridden in a class.

lilHermit replied:
What is the actual purpose of the const? It's only ever referenced by the Trait!

dereuromark replied:
That you can refer to it via class const instead of relying on magic strings.
This helps to avoid mistakes in spelling and eases migration/updates as the IDE knows about usage and what to rename.

robertpustulka replied:
You can as well use the trait in your class and define different event name using the const instead of rewriting the code.

Also [this line](https://github.com/cakephp/cakephp/blob/master/src/Validation/ValidatorAwareTrait.php#L174) needs to be changed to allow for late static binding.

lilHermit replied:
ok I'll let you guys decide where it goes but currently it breaks implementations where `Table` is not used. For example ([Type](https://github.com/cakephp/elastic-search/blob/master/src/Type.php))

lilHermit replied:
So what you're suggesting is (1) leave in `Table` (2) still have in `ValidatorAwareInterface` but change the value to something generic like `Event.buildValidator` and other non-`Table` implementations carry on as suggested and implement `ValidatorAwareInterface` or can override?

robertpustulka replied:
OK, so it needs to be fixed there. The class defines some validation related constants, but not the one that has been added recently.

Maybe the remaining constants also should be moved to the interface?

robertpustulka replied:
Yes, that's what I'm suggesting. Have default values in an interface and allow those to be overriden in classes.

lilHermit replied:
Yeah I agree all the validation related consts should be moved to the `ValidatorAwareInterface`

`BUILD_VALIDATOR_EVENT`, `VALIDATOR_PROVIDER_NAME` and `DEFAULT_VALIDATOR`

lilHermit committed:
Readded BUILD_VALIDATOR_EVENT back to table
ValidatorAwareInterface now has generic BUILD_VALIDATOR_EVENT and VALIDATOR_PROVIDER_NAME values

markstory commented on src/Validation/ValidatorAwareInterface.php:
```diff
@@ -25,7 +25,7 @@
      * @var string
      */
-    const BUILD_VALIDATOR_EVENT = 'Model.buildValidator';
+    const BUILD_VALIDATOR_EVENT = 'Event.buildValidator';
```
`Validation.buildValidator` might be a better name.

lilHermit committed:
Changed BUILD_VALIDATOR_EVENT const value

markstory commented:
Builds are failing with

```
PHP Fatal error:  Cannot inherit previously-inherited or override constant BUILD_VALIDATOR_EVENT from interface Cake\Validation\ValidatorAwareInterface in /home/travis/build/cakephp/cakephp/src/ORM/Table.php on line 127
```

robertpustulka commented:
Looks like interface constants cannot be overridden:

> It's possible for interfaces to have constants. Interface constants works exactly like class constants except they cannot be overridden by a class/interface that inherits them.
http://php.net/manual/en/language.oop5.interfaces.php

Sorry for the invalid suggestion, I wasn't aware of that.

How can we "fix" this and still allow for custom event names?

lilHermit commented:
Remove the constants and change to interface methods which return the string value?!? 

```
public function getBuildValidatorEvent();

public function getDefaultValidator();

public function getValidatorProviderName();
```

robertpustulka commented:
@lilHermit we can't add new methods to the interface I'm afraid due to the BC reasons.

Maybe we could check if a class constant is defined and fallback to the default value if it does not.

```php
$event = defined(static::BUILD_VALIDATOR_EVENT) ? static::BUILD_VALIDATOR_EVENT : 'Validation.buildValidator`;
```

lilHermit commented:
@robertpustulka 

Certainly an option, are we going to do the same for `DEFAULT_VALIDATOR` and `VALIDATOR_PROVIDER_NAME` where possible?

markstory commented:
The trait can use constants, and one of the requirements of  implementing the trait is that the implementer must provide those constants. We've had a few traits in the past that required methods to be implemented by the consuming class.

robertpustulka commented:
Maybe we just need to update the trait's docblock. The information about `BUILD_VALIDATOR_EVENT` [is missing there](https://github.com/lilHermit/cakephp-cakephp/blob/0b5155e46c10fc770f86b0972bf2adba095cf4e2/src/Validation/ValidatorAwareTrait.php#L28-L32).

markstory commented:
Additional constant documented in 61e19f3ba779d90bc84353a8c75fe488f05aa05c
